commit 50afde2f5afdd646ae2e395fe20684d9659f1c76
Author: Nigel Banks <nigel.g.banks@gmail.com>
Date:   Tue Apr 29 17:15:58 2025 +0200

    Fix for dynamically linked dependencies.

diff --git a/node.gyp b/node.gyp
index 33bc7d985f..a3243df609 100644
--- a/node.gyp
+++ b/node.gyp
@@ -537,6 +537,7 @@
       ],
 
       'dependencies': [
+        'deps/base64/base64.gyp:base64',
         'deps/histogram/histogram.gyp:histogram',
       ],
 
@@ -1227,6 +1228,7 @@
 
       'dependencies': [
         '<(node_lib_target_name)',
+        'deps/base64/base64.gyp:base64',
         'deps/histogram/histogram.gyp:histogram',
         'deps/ada/ada.gyp:ada',
       ],
@@ -1340,6 +1342,7 @@
 
       'dependencies': [
         '<(node_lib_target_name)',
+        'deps/base64/base64.gyp:base64',
         'deps/histogram/histogram.gyp:histogram',
         'deps/ada/ada.gyp:ada',
       ],
